pytorch_lightning/utilities/apply_func.py

```diff
@@ -34,6 +36,12 @@ def apply_to_collection(data: Any, dtype: Union[type, tuple], function: Callable
         return elem_type(*(apply_to_collection(d, dtype, function, *args, **kwargs) for d in data))
     elif isinstance(data, Sequence) and not isinstance(data, str):
         return elem_type([apply_to_collection(d, dtype, function, *args, **kwargs) for d in data])
+    elif isinstance(data, Batch):
+        new_batch = copy(data)  # Shallow copy is enough + I don't want to modify the object
+        for field in data.fields:
+            new_data = apply_to_collection(getattr(data, field), dtype, function, *args, **kwargs)
+            setattr(new_batch, field, new_data)
+        return new_batch
```

Comment:
I had a different approach in mind. I am not sure if the apply_to_collection is the right place to put this, since Batch is a terminal element and not a collection. Let me propose this:
We put the support into the method below, `move_data_to_device`:
we invoke `apply_to_collection `with 
```python
apply_to_collection(batch, dtype=(TransferableDataType, Batch), function=batch_to)
```
and extend the `to` closure with your code above.

Reply:
I think both ways are fine until the logic gets more complicated. You can also consider an object to be a collection (or rather composition) of simpler types. Anyway, I changed the code the way you proposed.

Reply:
> You can also consider an object to be a collection (or rather composition)

Yes, that was the motivation for adding a hook. If the Batch object had a "to" like the tensor does, it would make things much easier here.